railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,406 @@
+				if config, ok := plugin.Config["add"]; ok {
+					for _, header := range config.(map[string]interface{})["headers"].([]interface{}) {
+						if strings.HasPrefix(header.(string), "Knative-Serving-") {
+							log.Info("knative service updated. update knative CR condition and status...")
+							err := UpdateKnativeIngress(ctx, log, svc, kubeConfig, ips, hostname)
+							return fmt.Errorf("failed to update knative ingress err %v", err)
+						}
+					}
+				}
+			}
+		}
+
+		switch proto := *svc.Protocol; proto {
+		case "tcp":
+			err := UpdateTCPIngress(ctx, log, svc, kiccli, ips)
+			return fmt.Errorf("failed to update tcp ingress. err %v", err)
+		case "udp":
+			err := UpdateUDPIngress(ctx, log, svc, kiccli, ips)
+			return fmt.Errorf("failed to update udp ingress. err %v", err)
+		case "http":
+			err := UpdateIngressV1(ctx, log, svc, cli, ips)
+			return fmt.Errorf("failed to update ingressv1. err %v", err)
+		default:
+			log.Info("other 3rd party ingress not supported yet.")
+		}
+	}
+
+	return nil
+}
+
+func toKnativeLBStatus(coreLBStatus []apiv1.LoadBalancerIngress) []knative.LoadBalancerIngressStatus {
+	var res []knative.LoadBalancerIngressStatus
+	for _, status := range coreLBStatus {
+		res = append(res, knative.LoadBalancerIngressStatus{
+			IP:     status.IP,
+			Domain: status.Hostname,
+		})
+	}
+	return res
+}
+
+func retrieveNSAndNM(svc file.FService) (string, string, error) {
+	var name, namespace string
+	routeInf := strings.Split(*(svc.Routes[0].Name), ".")
+	namespace = routeInf[0]
+	name = routeInf[1]
+	if len(namespace) == 0 || len(name) == 0 {
+		return "", "", fmt.Errorf("configured route information is not completed which should not happen.")
+	}
+	return namespace, name, nil
+}
+
+// Update networking v1 ingress status
+func UpdateIngressV1(ctx context.Context, logger logr.Logger, svc file.FService, cli *clientset.Clientset,
+	ips []string) error {
+	namespace, name, err := retrieveNSAndNM(svc)
+	if err != nil {
+		return fmt.Errorf("failed to generate UDP client. err %v", err)
+	}
+
+	ingCli := cli.NetworkingV1().Ingresses(namespace)
+	cli.NetworkingV1beta1()
+	if err != nil {
+		return fmt.Errorf("failed to generate UDP client. err %v", err)
+	}
+	curIng, err := ingCli.Get(ctx, name, metav1.GetOptions{})
+	if err != nil || curIng == nil {
+		return fmt.Errorf("failed to fetch Ingress %v/%v: %w", namespace, name, err)
+	}
+
+	var status []apiv1.LoadBalancerIngress
+	sort.SliceStable(status, lessLoadBalancerIngress(status))
+	curIPs := curIng.Status.LoadBalancer.Ingress
+
+	status = SliceToStatus(ips)
+	if ingressSliceEqual(status, curIPs) {
+		log.Debugf("no change in status, update ingress v1 skipped")
+		return nil
+	}
+
+	curIng.Status.LoadBalancer.Ingress = status
+	retry := 0
+	for retry < status_update_retry {
+		_, err = ingCli.UpdateStatus(ctx, curIng, metav1.UpdateOptions{})
+		if err == nil {
+			break
+		}
+		log.Errorf("failed to update Ingress V1 status. %v. retrying...", err)
+		retry++
+	}
```

Comment:
This update mechanism has a problem: it does NOT properly retry in the event that the update fails due to a defunct object `ResourceVersion`. I recommend for these update loops we include the `curIng, err := ingCli.Get(ctx, name, metav1.GetOptions{})` within the loop which will make it so the retries are trying against the most recently found version of the resource.

Reply:
ingCli is a pointer parameter, so does curIng, the address is static, we just need update data whenever there is a failure. Don't really see the necessary fetching ingress each time.